pkg/providers/ironic.go

```diff
@@ -0,0 +1,200 @@
+package providers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/baremetal/httpbasic"
+	"github.com/gophercloud/gophercloud/openstack/baremetal/v1/nodes"
+	"github.com/gophercloud/gophercloud/pagination"
+	"go.etcd.io/etcd/pkg/transport"
+	"net/http"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"time"
+)
+
+var (
+	sometimeout = 30
+)
+
+type ironicProviderConfig struct {
+	Endpoint string `json:"endpoint"`
+	Username string `json:"username"`
+	Password string `json:"password"`
+	Image    string `json:"image"`
+}
+type ironicProvider struct {
+	config ironicProviderConfig
+	client *gophercloud.ServiceClient
+}
+
+func IronicProviderFactory(providerInfo string, secretData map[string][]byte) (Provider, error) {
+	config := ironicProviderConfig{
+		Username: "",
+		Password: "",
+		Endpoint: "https://172.22.0.3:6385",
+		Image:    "http://172.22.0.1/images/ofcir_image.qcow2",
+	}
+
+	if configJSON, ok := secretData["config"]; ok {
+		if err := json.Unmarshal(configJSON, &config); err != nil {
+			return nil, err
+		}
+	}
+
+	client, _ := httpbasic.NewBareMetalHTTPBasic(httpbasic.EndpointOpts{
+		IronicEndpoint:     config.Endpoint,
+		IronicUser:         config.Username,
+		IronicUserPassword: config.Password,
+	})
+
+	tlsInfo := transport.TLSInfo{
+		InsecureSkipVerify: true,
+	}
+	tlsTransport, _ := transport.NewTransport(tlsInfo, time.Second*30)
+	c := http.Client{
+		Transport: tlsTransport,
+	}
+
+	client.HTTPClient = c
+	client.Microversion = "1.74"
+
+	return &ironicProvider{
+		config: config,
+		client: client,
+	}, nil
+}
+
+func (p *ironicProvider) Acquire() (Resource, error) {
+	res := Resource{}
+
+	node, err := p.selectNode()
+	if err != nil {
+		return res, err
+	}
+
+	res.Id = node.UUID
+	return res, nil
+}
+
+func (p *ironicProvider) AcquireCompleted(id string) (bool, Resource, error) {
+	node, err := nodes.Get(p.client, id).Extract()
+
+	res := Resource{
+		Id: id,
+	}
+
+	if err != nil {
+		return false, res, err
+	}
+
+	if node.ProvisionState == "deploy failed" {
```

Comment:
What about using the ProvisionState const, such as https://github.com/gophercloud/gophercloud/blob/726767eb5c84ed7fc61e712c86feb9433a8b748a/openstack/baremetal/v1/nodes/requests.go#L28?